sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala

```diff
@@ -0,0 +1,107 @@
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.sources.{AlwaysTrue, CreatableRelationProvider, Filter, InsertableRelation}
+import org.apache.spark.sql.sources.v2.Table
+import org.apache.spark.sql.sources.v2.writer._
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+/**
+ * Physical plan node for append into a v2 table using V1 write interfaces.
+ *
+ * Rows in the output data set are appended.
+ */
+case class AppendDataExecV1(
+    writeBuilder: V1WriteBuilder,
+    plan: LogicalPlan) extends V1FallbackWriters {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    writeWithV1(writeBuilder.buildForV1Write())
+  }
+}
+
+/**
+ * Physical plan node for overwrite into a v2 table with V1 write interfaces. Note that when this
+ * interface is used, the atomicity of the operation depends solely on the target data source.
+ *
+ * Overwrites data in a table matched by a set of filters. Rows matching all of the filters will be
+ * deleted and rows in the output data set are appended.
+ *
+ * This plan is used to implement SaveMode.Overwrite. The behavior of SaveMode.Overwrite is to
+ * truncate the table -- delete all rows -- and append the output data set. This uses the filter
+ * AlwaysTrue to delete all rows.
+ */
+case class OverwriteByExpressionExecV1(
+    table: Table,
+    writeBuilder: V1WriteBuilder,
+    deleteWhere: Array[Filter],
+    plan: LogicalPlan) extends V1FallbackWriters {
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
+  }
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    writeBuilder match {
+      case builder: SupportsTruncate if isTruncate(deleteWhere) =>
+        writeWithV1(builder.truncate().asV1Writer.buildForV1Write())
+
+      case builder: SupportsOverwrite =>
+        writeWithV1(builder.overwrite(deleteWhere).asV1Writer.buildForV1Write())
+
+      case _ =>
+        throw new SparkException(s"Table does not support overwrite by expression: $table")
+    }
+  }
+}
+
+/** Some helper interfaces that use V2 write semantics through the V1 writer interface. */
+sealed trait V1FallbackWriters extends SupportsV1Write {
+  override def output: Seq[Attribute] = Nil
+  override final def children: Seq[SparkPlan] = Nil
+
+  protected implicit class toV1WriteBuilder(builder: WriteBuilder) {
+    def asV1Writer: V1WriteBuilder = builder match {
+      case v1: V1WriteBuilder => v1
+      case other => throw new IllegalStateException(
+        s"The returned writer ${other} was no longer a V1WriteBuilder.")
+    }
+  }
+}
+
+/**
+ * A trait that allows Tables that use V1 Writer interfaces to append data.
+ */
+trait SupportsV1Write extends SparkPlan {
+  def plan: LogicalPlan
+
+  protected def writeWithV1(relation: InsertableRelation): RDD[InternalRow] = {
+    relation.insert(Dataset.ofRows(sqlContext.sparkSession, plan), overwrite = false)
```

Comment:
Instead of passing the logical plan that will be run through the planner again, can we add a LogicalPlan that wraps a physical plan and a rule to convert that back to a physical plan?

Reply:
I think that's a great idea, and would be a useful node to have: `AlreadyPlanned` or something like that. I'd love to do that as a follow up.

Reply:
Sounds good to me.